src/functional/iterator/transforms/unroll_reduce.py

```diff
@@ -1,14 +1,35 @@
+from itertools import groupby
+from typing import Iterable
+
 from eve import NodeTranslator
 from eve.utils import UIDs
 from functional.iterator import ir
 
 
+def all_equal(it: Iterable):
+    g = groupby(it)
+    return next(g, True) and not next(g, False)
+
+
 class UnrollReduce(NodeTranslator):
     @staticmethod
-    def _get_last_offset(node: ir.Expr):
-        assert isinstance(node, ir.FunCall)
-        assert isinstance(node.fun, ir.FunCall) and node.fun.fun == ir.SymRef(id="shift")
-        return node.fun.args[0]
+    def _find_connectivity(reduce_args: Iterable[ir.Expr], offset_provider):
+        connectivities = []
+        for arg in reduce_args:
+            if (
+                isinstance(arg, ir.FunCall)
+                and isinstance(arg.fun, ir.FunCall)
+                and arg.fun.fun == ir.SymRef(id="shift")
+            ):
+                assert isinstance(arg.fun.args[0], ir.OffsetLiteral)
+                connectivities.append(offset_provider[arg.fun.args[0].value])
+
+        if not connectivities:
+            raise RuntimeError("Couldn't detect partial shift in any arguments of reduce.")
+
+        if not all_equal((c.max_neighbors, c.has_skip_values) for c in connectivities):
```

Comment:
…or maybe even just use the following which probably has the same readability for people used to Python:
```suggestion
        if len({(c.max_neighbors, c.has_skip_values) for c in connectivities}) != 1:
```